RD-5437 Redirect to the blueprint page after blueprint upload

leon-good-life opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->
<!--- e.g. "RD-1234 Improve styling in SuperComponent" -->
<!--- Note that the title is used as a commit message after merging the PR -->

## Description
Jira ticket: https://cloudifysource.atlassian.net/browse/RD-5197
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos
https://user-images.githubusercontent.com/17104212/182568002-a91993ca-36ac-471f-a6f5-d657dcc928fe.mov
<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
Latest `blueprints_spec` 2022-08-08 second time this day:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3398/pipeline

`blueprints_spec` 2022-08-08:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3397/pipeline

`blueprints_spec`:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3392/pipeline

`blueprints_spec`:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3331/pipeline

https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3322/pipeline
<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
Here is a Blueprints Widget Page, that has a description about blueprint upload. There is nothing to add:
https://docs.cloudify.co/latest/working_with/console/widgets/blueprints/

I'm not sure if adding "After the blueprint successfully uploaded the page is redirected to the blueprint page." is the right thing.

<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Check UI documentation guidelines - https://cloudifysource.atlassian.net/l/c/EYWJ0XfB - for details. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

kubama reviewed (dismissed):
No system test update is necessary?

qooban reviewed (changes requested):
Wait, it is stated that "User should be navigated to the uploaded blueprint page and be presented with the deployment modal" in RD-5437. 

Is deployment modal presented to the user? I guess not as there's no `openDeploymentModal` used as suggested in the ticket. 

System tests (for blueprints widget at least) would require an update I think. If not, I would add a test to check if the user gets redirected.

leon-good-life commented:
I addressed the review comment - open deployment modal: [56d63e9](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/56d63e92deacf2fded07a7724abddd476ab007ce)

https://user-images.githubusercontent.com/17104212/182579031-ecc3521b-dc88-45cd-9702-f8af618ea3d2.mov

leon-good-life commented:
I'm working on updating the `blueprints_spec`.

leon-good-life commented:
I fixed and started `blueprints_spec`: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3326/pipeline

leon-good-life commented:
I'm making changes and fixes in `blueprints_spec`.

leon-good-life commented:
I started `blueprint_spec`:

https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3331/pipeline

kubama commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -355,6 +357,9 @@ describe('Blueprints widget', () => {
                 cy.get('div[name=blueprintYamlFile] input').type(blueprintFileName);
                 cy.get('.button.ok').click();
 
+                cy.get('.deployBlueprintModal > .header').should('contain.text', 'Deploy blueprint');
+                cy.get('.deployBlueprintModal').within(() => cy.contains('Cancel').click());
+                cy.usePageMock('blueprints', blueprintsWidgetConfiguration).refreshTemplate();
```
Duplicated lines 346-348

leon-good-life replied:
I addressed it: [9e8477e](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/9e8477e8669a814a4bd4caecb9cd1407b234ce68)

kubama commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,7 +343,9 @@ describe('Blueprints widget', () => {
                 cy.contains('4/5: Parsing blueprint...');
-
+                cy.get('.deployBlueprintModal > .header').should('contain.text', 'Deploy blueprint');
+                cy.get('.deployBlueprintModal').within(() => cy.contains('Cancel').click());
```
```suggestion
                cy.get('.deployBlueprintModal').contains('Cancel').click();
```

leon-good-life replied:
I addressed it: [9e8477e](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/9e8477e8669a814a4bd4caecb9cd1407b234ce68)

kubama commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,7 +343,9 @@ describe('Blueprints widget', () => {
-
+                cy.get('.deployBlueprintModal > .header').should('contain.text', 'Deploy blueprint');
+                cy.get('.deployBlueprintModal').within(() => cy.contains('Cancel').click());
+                cy.usePageMock('blueprints', blueprintsWidgetConfiguration).refreshTemplate();
```
Why do we need to call `usePageMock` here?

leon-good-life replied:
Because I want to redirect back to blueprints. Maybe `cy.visit` is better? I can try.

kubama replied:
We already call `usePageMock` with same parameters in `beforeEach`, is it necessary to repeat it?

leon-good-life replied:
I will use "Blueprint Sources"  widget

leon-good-life replied:
I removed this line: [27d8149](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/27d81499edaaa968c6dfb0f492e4e780ab4d76e5)

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,7 +343,9 @@ describe('Blueprints widget', () => {
-
+                cy.get('.deployBlueprintModal > .header').should('contain.text', 'Deploy blueprint');
+                cy.get('.deployBlueprintModal').within(() => cy.contains('Cancel').click());
+                cy.usePageMock('blueprints', blueprintsWidgetConfiguration).refreshTemplate();
                 getBlueprintRow(blueprintName).contains('read-secret-blueprint.yaml');
```
If we want to check blueprint YAML file name, instead of getting back to the Blueprints page, maybe we can check it on the blueprint page (the one to which user was already redirected)? There's "Blueprint Sources" widget that keeps that information.

Same comment applies to the second test - L362-L363.

leon-good-life replied:
[27d8149](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/27d81499edaaa968c6dfb0f492e4e780ab4d76e5)

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,7 +343,9 @@ describe('Blueprints widget', () => {
                 cy.contains('3/5: Extracting blueprint...');
                 cy.contains('4/5: Parsing blueprint...');
-
+                cy.get('.deployBlueprintModal > .header').should('contain.text', 'Deploy blueprint');
+                cy.get('.deployBlueprintModal').within(() => cy.contains('Cancel').click());
```
Optional:

You possibly can have more elegant solution following Cypress best practices here (see https://docs.cypress.io/guides/references/best-practices#Selecting-Elements) if you try to minimize using CSS classes to get elements.

You can check the solution from `test/cypress/integration/widgets/blueprints_terraform_spec.ts` and have sth like:
```
cy.contains('Deploy blueprint')
    .parent('.modal')
    .within(() => cy.contains('Cancel').click())
```
WDYT?

leon-good-life replied:
I used your suggestion: [5b4bd27](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/5b4bd27031ea7d071ad425c887b457d45050343e)

leon-good-life commented:
I started `blueprints_spec`: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3392/pipeline

leon-good-life commented:
The test passed

kubama commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -311,7 +317,7 @@ describe('Blueprints widget', () => {
         });
 
-        describe('should upload a blueprint', () => {
+        describe.only('should upload a blueprint', () => {
```
`only` :slightly_smiling_face:

leon-good-life replied:
I removed it: c0c25a3

kubama reviewed (dismissed):
I think system test should be rerun after removing `only`. Other than that it looks good to me :+1: .

leon-good-life commented:
I started `blueprints_spec`: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3397/pipeline

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,8 +349,8 @@ describe('Blueprints widget', () => {
-
-                getBlueprintRow(blueprintName).contains('read-secret-blueprint.yaml');
+                closeDeployModal();
+                cy.get('[role="treeitem"]').contains('read-secret-blueprint.yaml');
```
The verification right now is less strict than previously as it doesn't verify if the YAML file is the main YAML file of that blueprint.

To narrow down the context and search only in Blueprint Sources widget, you can use `cy.getWidget('blueprintSources')`. To find main YAML file element in the list you probably can use something like `cy.contains('Main').parent()`.

leon-good-life replied:
I addressed it: c2491e9

leon-good-life commented:
I started system tests again: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3398/pipeline

I verified that they passing locally.

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -343,8 +349,8 @@ describe('Blueprints widget', () => {
-
-                getBlueprintRow(blueprintName).contains('read-secret-blueprint.yaml');
+                closeDeployModal();
+                cy.getWidget('blueprintSources').contains('Main').parent().contains('read-secret-blueprint.yaml');
```
It could also be extracted to a function as you are doing the same in L365.

leon-good-life replied:
I wonder how to call the function, maybe `blueprintSourceMainFile`?

leon-good-life replied:
[0dc295e](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/0dc295e31861d0672852a2a890f762c0e04bc0c9)

qooban reviewed (dismissed):
Thx for additional effort.

leon-good-life commented:
The test passed.

I run it again after addressing the last review comment:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3401/pipeline

qooban commented on test/cypress/integration/widgets/blueprints_spec.ts:
```diff
@@ -23,12 +23,22 @@ describe('Blueprints widget', () => {
         return cy.get(`#blueprintsTable_${blueprintName}`);
     }
 
+    function blueprintSourceMainFile() {
```
Function names are typically verbs or verb phrases, so I would go with something like `getBlueprintMainFile` or `getBlueprintYamlFile` or `getBlueprintMainYamlFile`.

leon-good-life replied:
I renamed it: [2924a43](https://github.com/cloudify-cosmo/cloudify-stage/pull/2117/commits/2924a43759b9e8617fdfaf878f4e9622ffc7dcd9)
